### PR TITLE
some C4 fixes in combinat folder

### PR DESCRIPTION
**Pull request by fchapoton**
Just fixing a few of the `ruff --select=C4` suggestions in the `combinat` folder

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.

### CODE REVIEW

File: src/sage/combinat/sloane_functions.py [hunk 1]
```diff
@@ -859,7 +859,7 @@ def _eval(self, n):
             sage: [sloane.A003418._eval(n) for n in range(1,11)]
             [1, 2, 6, 12, 60, 60, 420, 840, 2520, 2520]
         """
-        return arith.lcm([i for i in range(1, n+1)])
+        return arith.lcm(list(range(1, n + 1)))
```
Comment (mkoeppe) [inline src/sage/combinat/sloane_functions.py]: Shouldn't we rather fix the documentation of `sage.arith.functions.lcm` to clarify that the argument can be any iterable, not just a list or tuple?

File: src/sage/combinat/tiling.py [hunk 2]
```diff
@@ -1814,11 +1814,10 @@ def int_to_coord_dict(self):
...
         if self._reusable:
-            return dict((i, c) for i, c in enumerate(self.space()))
-        else:
-            number_of_pieces = len(self._pieces)
-            return dict((i+number_of_pieces, c)
-                        for i, c in enumerate(self.space()))
+            return dict(enumerate(self.space()))
+
+        number_of_pieces = len(self._pieces)
+        return {i + number_of_pieces: c for i, c in enumerate(self.space())}
```
Comment (mkoeppe) [inline src/sage/combinat/tiling.py]:
```suggestion
        return dict(enumerate(self.space(), start=number_of_pieces))
```

File: src/sage/combinat/words/morphism.py [hunk 3]
```diff
@@ -1727,9 +1729,8 @@ def is_uniform(self, k=None):
             True
         """
         if k is None:
-            return len(set(w.length() for w in self.images())) == 1
-        else:
-            return all(w.length() == k for w in self.images())
+            k = self.images()[0].length()
```
Comment (mkoeppe) [inline src/sage/combinat/words/morphism.py]: is there always at least one element?

Comment (fchapoton) [reply src/sage/combinat/words/morphism.py]: no, indeed

File: src/sage/combinat/words/morphism.py [hunk 4]
```diff
@@ -1725,11 +1727,19 @@ def is_uniform(self, k=None):
             False
             sage: tau.is_uniform(k=2)
             True
+
+        TESTS:
```
Comment (mkoeppe) [inline src/sage/combinat/words/morphism.py]:
```suggestion
        TESTS::
```

File: src/sage/combinat/words/morphism.py [hunk 5]
```diff
@@ -1725,11 +1727,19 @@ def is_uniform(self, k=None):
...
+
+            sage: phi = WordMorphism('')
+            sage: phi.is_uniform()
+            False
```
Comment (mkoeppe) [inline src/sage/combinat/words/morphism.py]: I know that this is the existing behavior, but is it correct? After all, we have
```
sage: phi.is_uniform(0)
True
```

File: src/sage/combinat/words/morphism.py [hunk 6]
```diff
@@ -1725,11 +1727,19 @@ def is_uniform(self, k=None):
...
         """
         if k is None:
-            return len(set(w.length() for w in self.images())) == 1
-        else:
-            return all(w.length() == k for w in self.images())
+            try:
+                k = self.images()[0].length()
+            except IndexError:
+                return False
```
Comment (mkoeppe) [inline src/sage/combinat/words/morphism.py]: Do we consider these 4 lines an improvement over the 1 line that was there before?

Taking into account my comment above, I would just suggest:
```
        if k is None:
            return len(set(w.length() for w in self.images())) <= 1
```

File: src/sage/combinat/words/morphism.py (hunk 5 above)
Comment (fchapoton) [reply src/sage/combinat/words/morphism.py]: indeed, I changed the answer to True

File: src/sage/combinat/words/morphism.py (hunk 6 above)
Comment (fchapoton) [reply src/sage/combinat/words/morphism.py]: This avoids to compute all of them in case of failure (similar to `all` and `any`).

Comment (mkoeppe) [reply src/sage/combinat/words/morphism.py]: Thanks, I missed that.

File: src/sage/combinat/words/morphism.py [hunk 7]
```diff
@@ -1725,11 +1728,19 @@ def is_uniform(self, k=None):
...
         """
         if k is None:
-            return len(set(w.length() for w in self.images())) == 1
-        else:
-            return all(w.length() == k for w in self.images())
+            try:
+                k = self.images()[0].length()
+            except IndexError:
+                return True
+        return all(w.length() == k for w in self.images())
```
Comment (mkoeppe) [inline src/sage/combinat/words/morphism.py:1743]: How about rewriting it using an iterator? Just bikeshedding, feel free to ignore.
```
images = iter(self.images())
if k is None:
    try:
        k = next(images).length()
    except StopIteration:
        return True
return all(w.length() == k for w in images)

**Review by mkoeppe — APPROVED** [review]

File: src/sage/combinat/words/morphism.py (hunk 7 above)
Comment (fchapoton) [reply src/sage/combinat/words/morphism.py:1743]: well, this is a list, so there is no point, I think

Comment (mkoeppe) [reply src/sage/combinat/words/morphism.py:1743]: Fine

**Pull request merged**